R/bs-dependencies.R

```diff
@@ -143,7 +143,12 @@ bs_theme_dependencies <- function(
         stylesheet = basename(out_file),
         script = basename(js_files),
         all_files = TRUE, # include font and map files
-        meta = list(viewport = "width=device-width, initial-scale=1, shrink-to-fit=no")
+        meta = list(viewport = "width=device-width, initial-scale=1, shrink-to-fit=no"),
+        # Enable busy indicators by default.
+        # This doesn't have much of anything to Bootstrap,
+        # but it's a convenient place to hang this feature's
+        # initialization code (so it works in Rmd etc).
+        head = if (isTRUE(version >= 5)) as.character(use_busy_indicators())
```

Comment:
This seems convenient but also greatly widens the scope of situations where the busy indicators dependencies will be included. I'm not sure I like that.

Similarly, the fact that `use_busy_indicators()` hoists `shiny::useBusyIndicators()` means that we're now directly connecting any Boostrap consumers to Shiny, which gives me pause. Additionally, this redirection makes it hard to reason about `use_busy_indicators()` because the implementation lives somewhere else entirely.

As a result of trying to figure out what `use_busy_indicators()` does, I realized that it should guard against `shiny` not being installed. `shiny` is a suggested dependency and `asNamespace()` will throw if it's not installed. And we wouldn't want to make access to Bootstrap via bslib dependent on the presence of `shiny` anyway.

Reply:
I'm less concerned about this after reading the implementation of [shiny::useBusyIndicators()](https://github.com/rstudio/shiny/blob/068b232e75347bac1a6aeef22e6611c6ae07cf0d/R/busy-indicators.R#L51-L70), at least from a technical perspective, but I don't like the interconnectedness this introduces.

At the very least, we should leave a comment here about why this is safe and we should have an additional comment over in shiny as a reminder of where this is called in bslib -- and that `useBusyIndicators()` can't change without considering how it's used in bslib.

Reply:
I hear you, there's definitely a set of tradeoffs I'm considering here. Here's a list of points that pushed me in this direction:

1. This approach is actually closer to how [busy indicators a brought into py-shiny](https://github.com/posit-dev/py-shiny/blob/dc0b2943b998d59216f6239d6a41d4fa8d754700/shiny/ui/_html_deps_py_shiny.py#L73-L81) 
2. We won't need to keep this logic in mind during future `page()`-related work.
3. By having the `<script>` in the `<head>` instead of the `<body>` we're reducing the possibility of other people's JS blocking our JS.
4. In the case of nested `page()` calls, this approach allows `useBusyIndicators()` to be placed _anywhere_ in the UI to disable/customize (Before this change, `useBusyIndicators()`  would need to be inside the inner-most page to win out).
5.  I think it's ultimately a good thing if we're more aggressive about enabling busy indication by default (and it seems lots of folks are still unfortunately using `fluidPage(theme = bs_theme())`). I'm slightly tempted to just say it's enabled by default for Shiny now (especially since I know that defaults work fine with `shinycssloaders::withSpinner()`), but that still seems like a step too far.

Reply:
How much of these tradeoffs would be resolved if spinners were enabled by default directly in Shiny? Either enabled by default in all cases or enabled by default for BS 5+?

To me, it seems cleaner to do this in Shiny; it seems like we're using bslib as a proxy to avoid breaking changes or being worried about being too aggressive with new features.

I'm very much in favor of handling all of this in Shiny and turning on busy indicators in the broadest sense we're comfortable with. That's most likely in Bootstrap 5 scenarios, but I'd be okay with turning it on in all of Shiny. (Or at the very least, seeing how many of our own tests would break as a result of turning it on everywhere.)

It's a good feature that brings improvements to all Shiny apps and I'd rather we make it broadly available than artificially limit it's reach.

Reply:
Yea, that would also give us 1-5, without the drawback of a more complicated user story / implementation. 

I feel on the fence about it -- we'll almost certainly cause some headaches where apps ultimately don't want to the busy indication (think Joe's cpuinfo PyShiny example), but the advice we can give is fairly straightforward.

Reply:
What if `fade` became officially supported by `useBusyIndicators()`? So that the fade could be turned off in the same way users can turn on or off the spinners?

Reply:
This idea has grown on me this morning when I realized I personally have a good use for this. In general, any app with automatic updates that doesn't want to broadcast that the server is doing work could use `fade = FALSE`. I also like that this would let shiny own both the fading as a busy indication and the way to turn it off.

Reply:
Yea, thanks for reminding me of that. After some more consideration, I think you're right (it's probably worth enabling by default and adding a `fade = FALSE` control). See https://github.com/rstudio/shiny/pull/4104. Will close here.